### PR TITLE
Themable menu

### PR DESCRIPTION
**Pull request by ThieryMichel**

### CODE REVIEW

File: config.json [hunk 1]
```diff
@@ -83,5 +83,82 @@
...
+      },
+      {
+        "label": {
+                "en": "About",
+                "fr": "A propos"
+        },
+        "icon": "/about.svg",
+        "position": "top",
+        "role": "custom",
+        "link": "/about.html"
```
Comment (Kmaschta) [inline config.json]: This page should not be used by default, right ?

File: src/api/controller/front.js [hunk 2]
```diff
@@ -146,9 +146,8 @@ export const getRenderingData = async (
 
 const handleRender = async (ctx, next) => {
     const { url, headers } = ctx.request;
-
     if (
-        url.match(/[^\\]*\.(\w+)$/) ||
+        (url.match(/[^\\]*\.(\w+)$/) && !url.match(/[^\\]*\.html$/)) ||
```
Comment (Kmaschta) [inline src/api/controller/front.js:152]: Why just not use `url.startsWith('/')` and `url.endsWith('.html')` .. ?

File: src/api/controller/index.js [hunk 3]
```diff
@@ -18,6 +21,28 @@ if (simulatedLatency) {
...
 app.use(mount('/api', api));
 
+const scriptRegEx = new RegExp('<script.*?( src=".*")?.*?>.*?</script>', 'gm');
+
+export const getScriptsFromHtml = html =>
+    (html.match(scriptRegEx) || [])
+        .map(script => {
+            const src = script.match(/<script.*?src="(.*?)".*?>/);
+            return src && src[1];
+        })
```
Comment (Kmaschta) [inline src/api/controller/index.js:31]: Please also check that the scripts are relative or on the same host.

File: src/app/custom/about.html [hunk 4]
```diff
@@ -0,0 +1,4 @@
+<div>
```
Comment (Kmaschta) [inline src/app/custom/about.html]: Are you sure you want to add this file by default ?

File: src/app/custom/about.js [hunk 5]
```diff
@@ -0,0 +1 @@
+alert('It works');
```
Comment (Kmaschta) [inline src/app/custom/about.js]: Hm.

File: src/app/js/public/MenuItem.js [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+    handleGraphItemClick,
+    toggleSearch,
+    canBeSearched,
+    role,
+    logout,
+}) => {
+    const label = config.label[polyglot.currentLocale];
+    const icon = getIcon(config.icon);
+
+    switch (config.role) {
```
Comment (Kmaschta) [inline src/app/js/public/MenuItem.js]: You should use `toLowerCase` here, just to be sure

File: src/app/js/public/MenuItem.js [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+                    activeClassName={css(styles.active)}
+                >
+                    {icon}
+                    {label}
+                </NavLink>
+            );
+        }
+        default:
+            console.error(
+                `Unknow role: ${config.role} menu item: ${JSON.stringify(
```
Comment (Kmaschta) [inline src/app/js/public/MenuItem.js]: You just have to use `custom` as the default, no ?
I would avoid to get an error message for nothing.

File: src/app/js/public/NavBar.js [hunk 8]
```diff
@@ -24,7 +14,15 @@ import Drawer from './Drawer';
 import Search from './search/Search';
 import GraphSummary from './graph/GraphSummary';
 import Favicon from './Favicon';
-import theme from '../theme';
+import jsonConfig from '../../../../config.json';
```
Comment (Kmaschta) [inline src/app/js/public/NavBar.js]: Are you importing the WHOLE config, with its secrets and stuff on the frontend build?

File: src/app/js/public/Routes.js [hunk 9]
```diff
@@ -8,6 +8,12 @@ import Resource from './resource/Resource';
 import Login from '../user/Login';
 import GraphPage from './graph/GraphPage';
 import NavBar from './NavBar';
+import jsonConfig from '../../../../config.json';
+import getCustomPage from './getCustomPage';
+
+const customRoutes = jsonConfig.front.menu.filter(
+    ({ role, link }) => (role !== 'custom' ? false : link.startsWith('/')),
```
Comment (Kmaschta) [inline src/app/js/public/Routes.js]: `custom` or no role at all?

File: src/app/js/public/getCustomPage.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+import Helmet from 'react-helmet';
+import translate from 'redux-polyglot/translate';
+import fetch from 'fetch-with-proxy';
+
+import { polyglot as polyglotPropTypes } from '../propTypes';
+
+export class CustomPage extends Component {
+    state = {};
+    UNSAFE_componentWillMount() {
+        fetch(`/customPage${this.props.link}`)
```
Comment (Kmaschta) [inline src/app/js/public/getCustomPage.js]: Why the `$` ? Why not just use a query param or just `/customPage/page`?

File: src/app/js/public/getCustomPage.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+            .catch(error => {
+                console.error(error);
+                this.setState({ error: true });
+            });
+    }
+    render() {
+        const { p: polyglot } = this.props;
+        const { html, scripts, error } = this.state;
+        if (error) {
+            return <div>{polyglot.t('page_not_found')}</div>;
```
Comment (Kmaschta) [inline src/app/js/public/getCustomPage.js]: Perhaps, try a redirect to the home ?
Just an idea.

File: src/app/js/public/getCustomPage.js [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        }
+        if (!html) {
+            return null;
```
Comment (Kmaschta) [inline src/app/js/public/getCustomPage.js]: Idem, error or redirect to the home

File: src/app/js/public/getCustomPage.js [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+CustomPage.propTypes = {
+    link: PropTypes.string.isRequired,
+    p: polyglotPropTypes.isRequired,
+};
+
+const getCustomPage = link =>
+    translate(function CustomPageForLink() {
+        return <CustomPage link={link} />;
```
Comment (Kmaschta) [inline src/app/js/public/getCustomPage.js]: This is not the best implementation that you can do.

**Review by Kmaschta — CHANGES_REQUESTED** [review]

File: src/app/js/public/getCustomPage.js (hunk 10 above)
Comment (ThieryMichel) [reply src/app/js/public/getCustomPage.js]: :trollface:

File: src/app/js/public/index.js [hunk 14]
```diff
@@ -30,16 +31,24 @@ const store = configureStore(
...
+        </MuiThemeProvider>
+    </Provider>,
     document.getElementById('root'),
 );
 
 // Hot Module Replacement API
 if (module.hot) {
     module.hot.accept('../Root', () => {
-        const NewRoot = require('../Root').default; // eslint-disable-line
+        const NewRoutes = require('./Routes').default; // eslint-disable-line
```
Comment (Kmaschta) [inline src/app/js/public/index.js]: Are you sure that we still need this `module.hot` condition now we have a webpack dev server?

File: src/app/translations.tsv [hunk 15]
```diff
@@ -423,3 +423,4 @@
 "host.isbn"	"ISBN"	"ISBN"
 "host.eisbn"	"E-ISBN"	"E-ISBN"
 "host.title"	"Title"	"Titre"
+"page_not_found"	"Page not found"	"Page non trouvée"
```
Comment (Kmaschta) [inline src/app/translations.tsv]: Is this translation still used.. ?

**Review by Kmaschta — CHANGES_REQUESTED** [review]

File: src/app/js/public/MenuItem.js (hunk 6 above)
Comment (ThieryMichel) [reply src/app/js/public/MenuItem.js]: nope, i think it's better to be strict for config

**Review by Kmaschta — APPROVED** [review]

**Pull request merged**